### PR TITLE
Removed deprecated usage of internal class ArrayChannel

### PR DESCRIPTION
**Pull request by gmariotti**
- Use an internal channel instead of a ReadStream to implement the ReceiveChannelHandler.
- Removed usage of internal ArrayChannel.

In progress @vietj:
- Unfortunately, I'm not able to figure it out what is the problem with the `HttpParserTest`, if I change line 68 from `chunk.length() - 2` to `chunk.length()` the test actually pass, but I don't understand why. I'd appreciate if you could help me there.
- I think that the change I made in the `ReadStream<T>.toChannel` resumes the stream immediately after a `receive/receiveOrNull/poll` is successful, but I'm not sure this is what you wanted.

### CODE REVIEW

**Comment by chengenzhao** [comment]
we got the same problem and 
seems like needs to add "an infinite number of items" 
https://groups.google.com/forum/?fromgroups#!topic/vertx-dev/B1kGCC5Wm-c
but no idea where to add it

**Comment by chengenzhao** [comment]
@gmariotti 
Hi, viet explained the reason in previous issue, and I just post comments in there, please check:
https://github.com/vert-x3/vertx-lang-kotlin/issues/87

**Comment by vietj** [comment]
I will review this soon thanks for submitting

**Comment by vietj** [comment]
your PR is actually very close, but there are two issues

- one in your code that does not pause the stream initially
- a few bugs in the record parser implementation

I will fix the bugs in the record parser implementation first

**Comment by gmariotti** [comment]
where should the stream be paused? In the test or in the implementation?

**Comment by vietj** [comment]
I've added print for debug and here is what I'm observing:

`Receive`/`Received` are printed by the actual HTTP parser test
`Send` is printed by `ReceiveChannelHandler#send` method

```
[vert.x-eventloop-thread-1 @coroutine#1] Receive line
[vert.x-eventloop-thread-1 @coroutine#2] Send to channel PUT /foo HTTP/1.1
[vert.x-eventloop-thread-1 @coroutine#3] Send to channel host: localhost:8080
[vert.x-eventloop-thread-1 @coroutine#4] Send to channel transfer-encoding: chunked
[vert.x-eventloop-thread-1 @coroutine#5] Send to channel 
[vert.x-eventloop-thread-1 @coroutine#6] Send to channel 6
[vert.x-eventloop-thread-1 @coroutine#7] Send to channel abc123
[vert.x-eventloop-thread-1 @coroutine#8] Send to channel 0
[vert.x-eventloop-thread-1 @coroutine#9] Send to channel 
[vert.x-eventloop-thread-1 @coroutine#1] Received PUT /foo HTTP/1.1
[vert.x-eventloop-thread-1 @coroutine#1] Receive header
[vert.x-eventloop-thread-1 @coroutine#1] Received host: localhost:8080
[vert.x-eventloop-thread-1 @coroutine#1] Receive header
[vert.x-eventloop-thread-1 @coroutine#1] Received transfer-encoding: chunked
[vert.x-eventloop-thread-1 @coroutine#1] Receive header
```

I can see that the coroutine is not resumed when we send the first element to the channel. To have this work we should have instead:

```
[vert.x-eventloop-thread-1 @coroutine#1] Receive line
[vert.x-eventloop-thread-1 @coroutine#2] Send to channel PUT /foo HTTP/1.1
[vert.x-eventloop-thread-1 @coroutine#1] Received PUT /foo HTTP/1.1
[vert.x-eventloop-thread-1 @coroutine#1] Receive header
[vert.x-eventloop-thread-1 @coroutine#3] Send to channel host: localhost:8080
[vert.x-eventloop-thread-1 @coroutine#1] Received host: localhost:8080
[vert.x-eventloop-thread-1 @coroutine#1] Receive header
[vert.x-eventloop-thread-1 @coroutine#4] Send to channel transfer-encoding: chunked
[vert.x-eventloop-thread-1 @coroutine#1] Received transfer-encoding: chunked
[vert.x-eventloop-thread-1 @coroutine#1] Receive header
[vert.x-eventloop-thread-1 @coroutine#5] Send to channel 
[vert.x-eventloop-thread-1 @coroutine#1] Received header
....
```

Any idea how we could achieve that ?

**Comment by vietj** [comment]
maybe we can discuss that on https://gitter.im/eclipse-vertx/vertx-users ?

**Comment by vietj** [comment]
it seems due to the channel capacity, i.e the following prints:

```
  runBlocking {
    val ch = Channel<String>(0)
    GlobalScope.launch {
      val s1 = ch.receive();
      println("${Thread.currentThread()} Received ${s1}")
      val s2 = ch.receive();
      println("${Thread.currentThread()} Received ${s2}")
    }
    println("${Thread.currentThread()} Sending a")
    ch.send("a")
    println("${Thread.currentThread()} Sent a")
    println("${Thread.currentThread()} Sending b")
    ch.send("b")
    println("${Thread.currentThread()} Sent b")
  }
```

```
Thread[main,5,main] Sending a
Thread[DefaultDispatcher-worker-1,5,main] Received a
Thread[main,5,main] Sent a
Thread[main,5,main] Sending b
Thread[DefaultDispatcher-worker-1,5,main] Received b
Thread[main,5,main] Sent b
```

and the following (changing the capacity 1):

```
Thread[main,5,main] Sending a
Thread[main,5,main] Sent a
Thread[main,5,main] Sending b
Thread[DefaultDispatcher-worker-1,5,main] Received a
Thread[DefaultDispatcher-worker-1,5,main] Received b
Thread[main,5,main] Sent b
```

and changing capacity to 2:

```
Thread[main,5,main] Sending a
Thread[main,5,main] Sent a
Thread[main,5,main] Sending b
Thread[main,5,main] Sent b
``

it does not print anything for some reason I ignore (and I'm not able to debug the source to understand the reason). So something is escaping my mind now :-)

**Comment by gmariotti** [comment]
It should work now, however, I have a doubt now:
- how to handle `receiveOrNull/poll/next`? The first 2 will fetch first, but if the channel was already closed or currently empty, they won't use the result. And a similar problem can happen with `next`.

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+  override fun offer(element: T): Boolean {
+    val isOffered = channel.offer(element)
+    if (isOffered) {
+      pauseStreamIfFull()
+    }
+    return isOffered
+  }
+
+  private fun pauseStreamIfFull() {
```
Comment (vietj) [inline vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt]: we should get rid of this function with the new fetch it's not necessary anymore

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    if (isFull) {
+      stream.pause()
+    }
+  }
+
+  override suspend fun receive(): T {
+    stream.fetch(1)
+    return channel.receive()
```
Comment (vietj) [inline vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt]: we should fetch after the receive operation

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+  }
+
+  override suspend fun receiveOrNull(): T? {
+    stream.fetch(1)
+    return channel.receiveOrNull()
```
Comment (vietj) [inline vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt]: we should fetch after the receive operation and only if it's not null

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  }
+
+  override fun poll(): T? {
+    stream.fetch(1)
+    return channel.poll()
```
Comment (vietj) [inline vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt]: we should fetch after the poll operation

File: vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+  override fun iterator(): ChannelIterator<T> {
+    return object : ChannelIterator<T> {
+      private val iterator = channel.iterator()
+      override suspend fun hasNext(): Boolean {
+        return iterator.hasNext()
+      }
+
+      override suspend fun next(): T {
+        stream.fetch(1)
+        return iterator.next()
```
Comment (vietj) [inline vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/ReceiveChannelHandler.kt]: fetch after next

**Comment by gmariotti** [comment]
@vietj those changes break again the `HttpParserTest.'test put chunked'` and the `ReceiveChannelHandlerTest.'test readStream to channel'`. I'll need to look into fixing those tests

**Comment by vietj** [comment]
@gmariotti ok let me know

**Comment by gmariotti** [comment]
@vietj I add the fixes based on your comments but I wasn't able to figure it out how to fix the `HttpParserTest` for the chunked case. IMHO, some kind of internal looking is probably needed. You want to start fetching and putting something inside the channel only when you actually have a `receive` request and not before. Not sure I'll be able to look into this PR before next week

**Comment by vietj** [comment]
@gmariotti can you push your code so I can have a look ?

**Comment by vietj** [comment]
one of the ReceivedChannelHandler test fails too

```
Running io.vertx.kotlin.coroutines.ReceiveChannelHandlerTest
Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.244 sec <<< FAILURE! - in io.vertx.kotlin.coroutines.ReceiveChannelHandlerTest
test readStream to channel(io.vertx.kotlin.coroutines.ReceiveChannelHandlerTest)  Time elapsed: 0.011 sec  <<< FAILURE!
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at io.vertx.kotlin.coroutines.ReceiveChannelHandlerTest.test readStream to channel(ReceiveChannelHandlerTest.kt:57)
```

**Comment by gmariotti** [comment]
Pushed already. The one in the ReceiveChannelHandler is related to the TestStream that is now not paused anymore

**Comment by vietj** [comment]
I will have a look in depth today. By chance do you know how I could debug the source code of coroutines with intellij ?

**Comment by gmariotti** [comment]
from IntelliJ, it should be possible to make Maven download the documentation and the source code of each library. And from there you should be able to put breakpoints in the coroutines library

**Comment by vietj** [comment]
do you have an idea where ?

**Comment by gmariotti** [comment]
Settings -> Build, Execution, Deployment -> Build Tools -> Maven -> Importing -> `Automatically download: Sources Documentation`

**Comment by vietj** [comment]
that won't work for me, but anywy I think I could solve the issue

**Comment by vietj** [comment]
I've merged this @gmariotti you can provide new PR for 1.3 migration

**Pull request merged**